retry.go

```diff
@@ -46,6 +46,16 @@ func (r *Retry) appendPostConditions(fns ...Condition) {
 	}
 }
 
+// preCondition is a function that decides whether the retry should continue.
+// It takes the previous error as an argument and returns an error
+// if the retry should stop. The error it returns is the error returned to the caller.
+// The reason it takes and returns an error is so that it can return an error describing
+// why the retry stopped. If there was a passed in error, instead of returning a brand new
+// error, it should wrap the error.
+// The passed in error will never be nil. Before the run function runs once, the passed in
+// error will be just be an error saying that the retry was not run even once.
+type preCondition func(error) error
```

Comment:
should just force the instructions in the doc by changing the signature to

`preCondition func(error) *conditionError`

and creating a

```go
type conditionError struct {
   condition string
   err error
}
```

Reply:
I thought about that, but then I decided to leave it like this for future flexibility. Its possible we want to return an error wrapped not with the `errors` package but with an actual wrapper type with some details on why the retry stopped. Its not something I see the use of at the moment, just something I figured could be nice in the future. If you don't think it could be useful, I'll change it.

Reply:
```go
type ConditionError interface {
    Condition() string
     error
}
```

?

Reply:
I don't want a consumer of the library to use string matching to figure out the condition that occurred. I was thinking a type for each condition.

Reply:
How about an interface + struct like:

```Go
type conditionError struct {
    origErr error
    retryErr error
}
// Error returns the original error, wrapped with a retry error message for context
func (c *conditionError) Error() string { return errors.Wrap(origErr, retryErr.Error()) }

// RetryError returns the retry error for equality tests and similar
func (c *conditionError) RetryError() error { return c.retryErr }

type ConditionError interface {
    Error() string
    RetryError() error
}
```

With that, callers can
- equality test their original error with `errors.Cause(err) == mylibrary.FunctionITriedtoRetry`
- equality test against retry errors with `err.RetryError() == retry.ErrWhatever`

Reply:
neither of these suggestions were implemented

Reply:
I responded in the main thread. We decided to defer this for now.